File: api/webhook.js
import axios from 'axios';
import crypto from 'crypto';
import { parseStringPromise } from 'xml2js';

const CORP_ID     = process.env.WXWORK_CORP_ID;
const AGENT_ID    = process.env.WXWORK_AGENT_ID;
const CORP_SECRET = process.env.WXWORK_SECRET;
const WX_TOKEN    = process.env.WXWORK_TOKEN;
const WX_AES_KEY  = Buffer.from(process.env.WXWORK_AES_KEY + '=', 'base64');

const SUPABASE_URL    = process.env.SUPABASE_URL;
const SUPABASE_KEY    = process.env.SUPABASE_KEY;
const SILICONFLOW_KEY = process.env.SILICONFLOW_KEY;

// ── 企业微信签名验证 ──────────────────────────────────
function verifySignature(signature, timestamp, nonce, data = '') {
  const items = [WX_TOKEN, timestamp, nonce, data].sort().join('');
  const hash = crypto.createHash('sha1').update(items).digest('hex');
  return hash === signature;
}

// ── AES 解密 ─────────────────────────────────────────
function decrypt(encrypted) {
  const buf = Buffer.from(encrypted, 'base64');
  const decipher = crypto.createDecipheriv('aes-256-cbc', WX_AES_KEY, WX_AES_KEY.slice(0, 16));
  decipher.setAutoPadding(false);
  const decrypted = Buffer.concat([decipher.update(buf), decipher.final()]);
  // 去掉 PKCS7 padding
  const pad = decrypted[decrypted.length - 1];
  const content = decrypted.slice(16, decrypted.length - pad);
  const msgLen = content.readUInt32BE(0);
  return content.slice(4, 4 + msgLen).toString('utf-8');
}

// ── AES 加密 ─────────────────────────────────────────
function encrypt(msg) {
  const random = crypto.randomBytes(16);
  const msgBuf = Buffer.from(msg, 'utf-8');
  const lenBuf = Buffer.alloc(4);
  lenBuf.writeUInt32BE(msgBuf.length);
  const corpBuf = Buffer.from(CORP_ID, 'utf-8');
  let plain = Buffer.concat([random, lenBuf, msgBuf, corpBuf]);
  // PKCS7 padding
  const pad = 32 - (plain.length % 32);
  const padBuf = Buffer.alloc(pad, pad);
  plain = Buffer.concat([plain, padBuf]);
  const cipher = crypto.createCipheriv('aes-256-cbc', WX_AES_KEY, WX_AES_KEY.slice(0, 16));
  cipher.setAutoPadding(false);
  return Buffer.concat([cipher.update(plain), cipher.final()]).toString('base64');
}

// ── 获取 access_token ────────────────────────────────
async function getAccessToken() {
  const res = await axios.get(
    `https://qyapi.weixin.qq.com/cgi-bin/gettoken?corpid=${CORP_ID}&corpsecret=${CORP_SECRET}`
  );
  return res.data.access_token;
}

// ── 主动推送消息 ──────────────────────────────────────
async function sendMessage(toUser, content) {
  const token = await getAccessToken();
  await axios.post(
    `https://qyapi.weixin.qq.com/cgi-bin/message/send?access_token=${token}`,
    { touser: toUser, msgtype: 'text', agentid: parseInt(AGENT_ID), text: { content } }
  );
}

// ── Embedding ─────────────────────────────────────────
async function getEmbedding(text) {
  const res = await axios.post(
    'https://api.siliconflow.cn/v1/embeddings',
    { model: 'BAAI/bge-m3', input: text.slice(0, 2000), encoding_format: 'float' },
    { headers: { Authorization: `Bearer ${SILICONFLOW_KEY}` } }
  );
  return res.data.data[0].embedding;
}

// ── 向量检索 ──────────────────────────────────────────
async function searchDocuments(embedding) {
  const res = await axios.post(
    `${SUPABASE_URL}/rest/v1/rpc/match_documents`,
    { query_embedding: embedding, match_count: 5, match_threshold: 0.5 },
    { headers: { apikey: SUPABASE_KEY, Authorization: `Bearer ${SUPABASE_KEY}` } }
  );
  return res.data;
}

// ── LLM 生成 ──────────────────────────────────────────
async function generateAnswer(question, contexts) {
  if (!contexts || contexts.length === 0) {
    return '我在知识库中没有找到相关内容，请尝试换个问法。';
  }
  const contextText = contexts
    .map(c => `【来源：${c.file_path}】\n${c.content}`)
    .join('\n\n---\n\n');

  const prompt = `你是吹雨的个人知识库助手，帮助主人检索和整理他的知识笔记。

下面是从知识库中检索到的相关内容：

${contextText}

---

请根据以上知识库内容，回答主人的问题。要求：
1. 回答要简洁、准确，直接基于知识库内容
2. 如果知识库内容不足以完整回答，请说明哪些部分有记录、哪些没有
3. 适当指出内容来源
4. 用中文回答

主人的问题：${question}`;

  const res = await axios.post(
    'https://api.siliconflow.cn/v1/chat/completions',
    {
      model: 'deepseek-ai/DeepSeek-V3',
      messages: [{ role: 'user', content: prompt }],
      max_tokens: 1024,
      temperature: 0.3,
    },
    { headers: { Authorization: `Bearer ${SILICONFLOW_KEY}` } }
  );
  return res.data.choices[0].message.content;
}

// ── RAG 主流程 ────────────────────────────────────────
async function rag(question) {
  const embedding = await getEmbedding(question);
  const contexts  = await searchDocuments(embedding);
  return generateAnswer(question, contexts);
}

// ── Vercel Edge Handler ───────────────────────────────
export const config = { runtime: 'edge' };

export default async function handler(req) {
  const url    = new URL(req.url);
  const params = url.searchParams;

  // GET：企业微信验证
  if (req.method === 'GET') {
    const msgSignature = params.get('msg_signature') || '';
    const timestamp    = params.get('timestamp') || '';
    const nonce        = params.get('nonce') || '';
    const echostr      = params.get('echostr') || '';

    if (verifySignature(msgSignature, timestamp, nonce, echostr)) {
      const decrypted = decrypt(echostr);
      return new Response(decrypted, { status: 200 });
    }
    return new Response('Forbidden', { status: 403 });
  }

  // POST：接收消息
  if (req.method === 'POST') {
    const msgSignature = params.get('msg_signature') || '';
    const timestamp    = params.get('timestamp') || '';
    const nonce        = params.get('nonce') || '';

    const body = await req.text();
    let encrypted;
    try {
      const parsed = await parseStringPromise(body);
      encrypted = parsed.xml.Encrypt[0];
    } catch {
      return new Response('Bad Request', { status: 400 });
    }

    if (!verifySignature(msgSignature, timestamp, nonce, encrypted)) {
      return new Response('Forbidden', { status: 403 });
    }

    const xmlStr  = decrypt(encrypted);
    const parsed  = await parseStringPromise(xmlStr);
    const msgType = parsed.xml.MsgType?.[0];
    const userId  = parsed.xml.FromUserName?.[0];
    const content = parsed.xml.Content?.[0]?.trim();

    if (msgType === 'text' && userId && content) {
      // waitUntil 让后台任务在响应返回后继续执行
      const ctx = req.waitUntil
        ? req
        : { waitUntil: (p) => p };

      ctx.waitUntil(
        rag(content)
          .then(answer => sendMessage(userId, answer))
          .catch(err => sendMessage(userId, `处理出错：${err.message}`))
      );
    }

    return new Response('success', { status: 200 });
  }

  return new Response('Method Not Allowed', { status: 405 });
}

File: api/webhook.py
"""
webhook.py - 异步处理版本
收到消息立即返回200，后台线程处理RAG并主动推送结果
"""

import os
import sys
import time
import threading
import xml.etree.ElementTree as ET
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs

sys.path.insert(0, os.path.dirname(__file__))
from wxwork import WXBizMsgCrypt
from rag import query as rag_query

CORP_ID     = os.environ.get("WXWORK_CORP_ID", "")
AGENT_ID    = os.environ.get("WXWORK_AGENT_ID", "")
CORP_SECRET = os.environ.get("WXWORK_SECRET", "")
WX_TOKEN    = os.environ.get("WXWORK_TOKEN", "")
WX_AES_KEY  = os.environ.get("WXWORK_AES_KEY", "")

crypt = WXBizMsgCrypt(WX_TOKEN, WX_AES_KEY, CORP_ID)


def get_access_token() -> str:
    import requests
    url = f"https://qyapi.weixin.qq.com/cgi-bin/gettoken?corpid={CORP_ID}&corpsecret={CORP_SECRET}"
    resp = requests.get(url, timeout=10)
    return resp.json().get("access_token", "")


def send_message(to_user: str, content: str):
    import requests
    token = get_access_token()
    if not token:
        return
    url = f"https://qyapi.weixin.qq.com/cgi-bin/message/send?access_token={token}"
    payload = {
        "touser": to_user,
        "msgtype": "text",
        "agentid": int(AGENT_ID),
        "text": {"content": content},
    }
    requests.post(url, json=payload, timeout=30)


def process_and_reply(user_id: str, question: str):
    """后台线程：执行RAG查询并主动推送结果"""
    try:
        answer = rag_query(question)
        send_message(user_id, answer)
    except Exception as e:
        send_message(user_id, f"处理出错：{str(e)}")


def parse_xml(xml_str: str) -> dict:
    root = ET.fromstring(xml_str)
    return {child.tag: child.text for child in root}


class handler(BaseHTTPRequestHandler):

    def do_GET(self):
        params = parse_qs(urlparse(self.path).query)
        msg_signature = params.get("msg_signature", [""])[0]
        timestamp     = params.get("timestamp", [""])[0]
        nonce         = params.get("nonce", [""])[0]
        echostr       = params.get("echostr", [""])[0]

        if crypt.verify_signature(msg_signature, timestamp, nonce, echostr):
            decrypted = crypt.decrypt(echostr)
            self._respond(200, decrypted)
        else:
            self._respond(403, "Forbidden")

    def do_POST(self):
        params = parse_qs(urlparse(self.path).query)
        msg_signature = params.get("msg_signature", [""])[0]
        timestamp     = params.get("timestamp", [""])[0]
        nonce         = params.get("nonce", [""])[0]

        length = int(self.headers.get("Content-Length", 0))
        body   = self.rfile.read(length).decode("utf-8")

        try:
            root      = ET.fromstring(body)
            encrypted = root.find("Encrypt").text
        except Exception:
            self._respond(400, "Bad Request")
            return

        if not crypt.verify_signature(msg_signature, timestamp, nonce, encrypted):
            self._respond(403, "Forbidden")
            return

        xml_str = crypt.decrypt(encrypted)
        msg     = parse_xml(xml_str)

        if msg.get("MsgType") == "text":
            user_id  = msg.get("FromUserName", "")
            question = msg.get("Content", "").strip()

            # 立即返回200，避免微信超时重试
            self._respond(200, "success")

            # 后台线程处理RAG并推送结果
            t = threading.Thread(target=process_and_reply, args=(user_id, question))
            t.daemon = True
            t.start()
        else:
            self._respond(200, "success")

    def _respond(self, code: int, body: str):
        self.send_response(code)
        self.send_header("Content-Type", "text/plain")
        self.end_headers()
        self.wfile.write(body.encode("utf-8"))

    def log_message(self, format, *args):
        pass